File: Python/project_euler/problem_04.py
"""https://projecteuler.net/problem=4"""


def is_palindrome(integer: int) -> bool:
    """Checks if the integer is a palindrome.

    Arguments:
        integer {int} -- The integer to check.

    Returns:
        bool -- True if the integer is a palindrome otherwise False.
    """
    return str(integer) == reverse_integer(integer)


def reverse_integer(integer: int) -> str:
    """Reverses the order of the digits in an integer.

    Arguments:
        integer {int} -- The integer to be reversed.

    Returns:
        str -- A string of the reversed integer.
    """
    return str(integer)[::-1]


def solve() -> int:
    """Solves problem 4.

    Uses a nested for-loop to iterate from 999 to 100 and checks whether the product
    is a palindrome. The inner loop will not iterate to 99; it will iterate to the
    current value of the parent loop.

    Returns:
        int -- The largest product that is a palindrome.
    """
    largest_palindrome = 0
    for i in range(999, 99, -1):
        for j in range(999, i, -1):
            product = i * j
            if product <= largest_palindrome:
                break
            if is_palindrome(product):
                largest_palindrome = product
    return largest_palindrome

File: C#/Miller.ProjectEuler.Solutions/Problem04.cs
using System.Text;

namespace Miller.ProjectEuler.Solutions
{
    /// <summary>
    /// https://projecteuler.net/problem=4
    /// </summary>
    public sealed class Problem04 : IProblem
    {
        public string Name => "Largest palindrome product";

        /// <summary>
        /// Finds the largest palindrome from the product of two 3-digit numbers.
        /// </summary>
        /// <returns>Returns 906609</returns>
        public string Solve()
        {
            var largestPalindrome = 0;
            for (int i = 100; i <= 999; i++)
            {
                for (int j = 100; j < 999; j++)
                {
                    var product = i * j;
                    if (product > largestPalindrome && IsPalindrome(product.ToString()))
                    {
                        largestPalindrome = product;
                    }
                }
            }
            return largestPalindrome.ToString();
        }

        /// <summary>
        /// Evaluates whether the string is a palindrome by comparing the string against its reverse.
        /// </summary>
        /// <param name="str">The string to evaluate.</param>
        /// <returns>True if the string is a palindrome and false otherwise.</returns>
        private bool IsPalindrome(string str)
        {
            var builder = new StringBuilder();
            for (int i = str.Length - 1; i >= 0; i--)
            {
                builder.Append(str[i]);
            }
            var reversed = builder.ToString();
            return str == reversed;
        }
    }
}